kernel/aster-nix/src/thread/mod.rs

```diff
@@ -67,23 +67,36 @@ impl Thread {
 
     /// Run this thread at once.
     pub fn run(&self) {
-        self.status.lock().set_running();
+        self.set_status(ThreadStatus::Running);
         self.task.run();
     }
 
     pub fn exit(&self) {
-        let mut status = self.status.lock();
-        if !status.is_exited() {
-            status.set_exited();
-        }
+        self.set_status(ThreadStatus::Exited);
+    }
+
+    /// Returns the current status.
+    pub fn status(&self) -> ThreadStatus {
+        ThreadStatus::try_from(self.status.load(Ordering::Acquire)).unwrap()
     }
 
-    pub fn is_exited(&self) -> bool {
-        self.status.lock().is_exited()
+    /// Updates the status with the `new` value.
+    pub fn set_status(&self, new: ThreadStatus) {
+        self.status.store(new as u8, Ordering::Release);
     }
 
-    pub fn status(&self) -> &Mutex<ThreadStatus> {
-        &self.status
+    /// Sets a `new` value to the status if the current status is the same as the `current` value.
+    ///
+    /// Returns a boolean indicating whether the new value was written and containing the current value.
+    pub fn compare_exchange_status(&self, current: ThreadStatus, new: ThreadStatus) -> bool {
+        self.status
+            .compare_exchange(
+                current as u8,
+                new as u8,
+                Ordering::Acquire,
+                Ordering::Relaxed,
```

Comment:
I think that you mean `Ordering::Release`, although I'm not sure whether or not this memory order is necessary for the thread status.

Reply:
The `success` describes the required ordering for the read-modify-write operation that takes place if the comparison with `current` succeeds.

I think `AcqRel` is correct, it has the effects of both Acquire and Release together: For loads it uses Acquire ordering. For stores it uses the Release ordering.

Reply:
Yes, but you are writing `Order::Relaxed`. So you should write `Order::Release` instead. I mean, there is a typo.

Reply:
Sorry, I misunderstood your comment, so it seems we should write `success = Ordering::AcqRel`, `failure = Ordering::Relaxed/Acquire` (the `failure` order depends on whether we want to synchronize with the last status write if the exchange fails).